server/lease.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pkg/errors"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+// LeaderLease is used for renewing leadership of PD server.
+type LeaderLease struct {
+	client       *clientv3.Client
+	lease        clientv3.Lease
+	ID           clientv3.LeaseID
+	leaseTimeout time.Duration
+
+	expireTime atomic.Value
+}
+
+// NewLeaderLease creates a lease.
+func NewLeaderLease(client *clientv3.Client) *LeaderLease {
+	return &LeaderLease{
+		client: client,
+		lease:  clientv3.NewLease(client),
+	}
+}
+
+// Grant uses `lease.Grant` to initialize the lease and expireTime.
+func (l *LeaderLease) Grant(leaseTimeout int64) error {
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(l.client.Ctx(), requestTimeout)
+	leaseResp, err := l.lease.Grant(ctx, leaseTimeout)
+	cancel()
+	if cost := time.Since(start); cost > slowRequestTime {
+		log.Warn("lease grants too slow", zap.Duration("cost", cost))
+	}
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	l.ID = leaseResp.ID
+	l.leaseTimeout = time.Duration(leaseTimeout) * time.Second
+	l.expireTime.Store(start.Add(time.Duration(leaseResp.TTL) * time.Second))
+	return nil
+}
+
+// Close releases the lease.
+func (l *LeaderLease) Close() error {
+	return l.lease.Close()
+}
+
+// IsExpired checks if the lease is expired. If it returns true, current PD
+// server should step down and try to re-elect again.
+func (l *LeaderLease) IsExpired() bool {
+	return time.Now().After(l.expireTime.Load().(time.Time))
+}
+
+// KeepAlive auto renews the lease and update expireTime.
+func (l *LeaderLease) KeepAlive(ctx context.Context) {
```

Comment:
I found that the main enhancement to maintain the `expireTime`, so why not use etcd's `KeepAlive` and update `expireTime` when got a response from the keepalive channel?

Reply:
> I found that the main enhancement is to maintain the `expireTime`, so why not use etcd's `KeepAlive` and update `expireTime` when got a response from the keepalive channel?

I think the problem here is that `KeepAlive` only returns TTL. We cannot know the "start time" of the TTL (`time.Now()` is not the answer), so we cannot calculate the `expireTime`.

Reply:
OK, thanks